addons/flannel/0.20.2/install.sh

```diff
@@ -33,6 +33,45 @@ function flannel_pre_init() {
     fi
 
     flannel_init_pod_subnet
+
+    if ! flannel_weave_conflict; then
+        return 0
+    fi
+
+    # the code below deals with weave to flannel migration and is not needed for new installs.
+    logWarn "The migration from Weave to Flannel will require whole-cluster downtime."
+    logWarn "Would you like to continue?"
+    if ! confirmY ; then
+        bail "Not migrating from Weave to Flannel"
+    fi
+
+    local node_count
+    node_count="$(kubectl get nodes --no-headers 2>/dev/null | wc -l)"
+    if [ "$node_count" = "1" ]; then
+        return 0
+    fi
+
+    local kurl_util_image
+    kurl_util_image=$(crictl images 2>/dev/null | grep replicated/kurl-util | sort | tail -1 | awk '//{printf "%s:%s\n",$1,$2;}')
+    if [ -z "$kurl_util_image" ]; then
+        if [ "$AIRGAP" = "1" ]; then
+            logFail "Your airgap installation misses replicated/kurl-util image, please contact support."
+            bail "Not migrating from Weave to Flannel"
+        fi
+        # at this point we haven't found the kurl-util image in the local store, but as this is
+        # not an airgap installation we can still try to pull it.
+        kurl_util_image="replicated/kurl-util:latest"
+        if [ -n "$KURL_VERSION" ]; then
+            kurl_util_image="replicated/kurl-util:$KURL_VERSION"
+        fi
+    fi
+
```

Comment:
I may have misunderstood the code but `${K8S_DISTRO}_addon_for_each addon_pre_init` is called before `install_cri` so I don't think we can possibly have loaded any image before running add-ons pre_init. Isn't that correct ?

Reply:
For sake of completeness (and assuming my understanding as correct) we are attempting to use, during the upgrade, an image that was loaded during the install.

Reply:
You are correct. Only the remote nodes should have the image as we prompt for the task to load images in common_prompts.

As this only runs on upgrades and the cri is already installed, I think it would be better then to call get_shared and load the kurl-util image one additional time. Something like below.

```suggestion

    if [ "$AIRGAP" = "1" ]; then
        # load the KURL_UTIL_IMAGE as the nodes connectivity check relies on it
        # this image will be pulled in online installs and loaded by the load-images task for airgap on remote nodes
        get_shared
    fi

```

Reply:
I was wondering how the `load-images` would work in an air-gapped environment when migrating from Docker to Containerd. The [code ](https://github.com/replicatedhq/kURL/blob/main/scripts/tasks.sh#L139-L149) seems to use `ctr` whenever `DOCKER_VERSION` isn't set.  I realise now that we install `containerd` whenever we install docker, so I guess that is why it  works.

Reply:
I hadn't thought about that. Did you test it and does it work?

Reply:
Yes, it does work, and it has nothing to do with `ctr` as I initially thought. `DOCKER_VERSION`, on tasks.sh, is defined [here](https://github.com/replicatedhq/kURL/blob/main/scripts/tasks.sh#L433-L438). I was expecting it to be the YAML value but it is not.

Reply:
OK, it seems like `load-images` is not loading kurl-util image:

```
$ cat ./tasks.sh | sudo bash -s load-images
⚙  Running tasks with the argument(s): load-images
loading all infrastructure images
Loaded image: minio/minio:RELEASE.2023-01-25T00-19-54Z
Loaded image: registry.k8s.io/pause:3.6
Loaded image: k8s.gcr.io/pause:3.5
Loaded image: registry.k8s.io/pause:3.6
Loaded image: k8s.gcr.io/pause:3.1
Loaded image: k8s.gcr.io/pause:3.2
Loaded image: registry.k8s.io/pause:3.6
Loaded image: k8s.gcr.io/pause:3.2
Loaded image: openebs/provisioner-localpv:3.3.0
Loaded image: openebs/node-disk-operator:2.0.0
Loaded image: openebs/linux-utils:3.3.0
Loaded image: openebs/node-disk-manager:2.0.0
Loaded image: kurlsh/s3cmd:20221219-ff60227
Loaded image: registry:2.8.1
Loaded image: flannel/flannel:v0.21.3
Loaded image: flannel/flannel-cni-plugin:v1.1.2
Loaded image: flannel/flannel:v0.21.4
Loaded image: flannel/flannel-cni-plugin:v1.1.2
Loaded image: k8s.gcr.io/etcd:3.4.13-0
Loaded image: k8s.gcr.io/pause:3.2
Loaded image: k8s.gcr.io/coredns:1.7.0
Loaded image: k8s.gcr.io/kube-proxy:v1.19.16
Loaded image: k8s.gcr.io/kube-scheduler:v1.19.16
Loaded image: k8s.gcr.io/kube-controller-manager:v1.19.16
Loaded image: k8s.gcr.io/kube-apiserver:v1.19.16
Loaded image: registry.k8s.io/etcd:3.5.6-0
Loaded image: registry.k8s.io/pause:3.9
Loaded image: registry.k8s.io/coredns/coredns:v1.9.3
Loaded image: registry.k8s.io/kube-proxy:v1.26.3
Loaded image: registry.k8s.io/kube-scheduler:v1.26.3
Loaded image: registry.k8s.io/kube-controller-manager:v1.26.3
Loaded image: registry.k8s.io/kube-apiserver:v1.26.3
Loaded image: registry.k8s.io/etcd:3.5.6-0
Loaded image: registry.k8s.io/pause:3.8
Loaded image: registry.k8s.io/coredns/coredns:v1.9.3
Loaded image: registry.k8s.io/kube-proxy:v1.25.8
Loaded image: registry.k8s.io/kube-scheduler:v1.25.8
Loaded image: registry.k8s.io/kube-controller-manager:v1.25.8
Loaded image: registry.k8s.io/kube-apiserver:v1.25.8
Loaded image: registry.k8s.io/etcd:3.5.6-0
Loaded image: registry.k8s.io/pause:3.7
Loaded image: registry.k8s.io/coredns/coredns:v1.8.6
Loaded image: registry.k8s.io/kube-proxy:v1.24.12
Loaded image: registry.k8s.io/kube-scheduler:v1.24.12
Loaded image: registry.k8s.io/kube-controller-manager:v1.24.12
Loaded image: registry.k8s.io/kube-apiserver:v1.24.12
```

I have manually deleted the kurl-util image before the upgrade.

Reply:
Oh, it seems like I don't have the file `shared/kurl-util.tar` after the watchrync.

Reply:
`make build/shared` seems to have addressed the missing image, now I have the `shared/kurl-util.tar` file.

Reply:
@emosbaugh Tested in an air-gapped environment, it works.